### PR TITLE
[clang] Add test for CWG2811 "Clarify "use" of main"

### PR DESCRIPTION
**Pull request by Endilll**
This patch covers [CWG2811](https://cplusplus.github.io/CWG/issues/2811.html) "Clarify "use" of main", basically adding a test for `-Wmain`, focusing on usages of `main` in unevaluated contexts.

To my understanding, the diagnostic message is based on the wording, so I updated it based on the new wording. I also replaces "ISO C++ requires" with a phrasing that explicitly says "extension".

### CODE REVIEW

**Comment by cor3ntin** [comment]
I think I prefer the current wording, it's clearer that it's not conforming (at the very least we should say something like "clang extension". and maybe "referring to main is a clang extension".
LGTM otherwise

do we have test that this is fine?

```
namespace foo {
    int main();
    using a = decltype(&main);
}```

**Comment by Endilll** [comment]
> it's clearer that it's not conforming (at the very least we should say something like "clang extension"

As someone who was a user of Clang not so long ago, I have to say that the only thing "ISO C++ requires" conveyed to me was "you're doing something wrong". For me it never had a clear interpretation "you're asking for an extension, but we're not giving it to you". On the other hand, I believe the word "extension" is clearly understood to describe something outside of ISO C++, somewhat by definition. At least that would match my understanding 3-4 years ago.

CC @AaronBallman I remember we discussed it offline a while ago.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 1]
```diff
@@ -964,7 +964,7 @@ def err_main_global_variable :
 def warn_main_redefined : Warning<"variable named 'main' with external linkage "
     "has undefined behavior">, InGroup<Main>;
 def ext_main_used : Extension<
-  "ISO C++ does not allow 'main' to be used by a program">, InGroup<Main>;
+  "expressions that refer to 'main' are an extension">, InGroup<Main>;
```
Comment (AaronBallman) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:967]:
```suggestion
  "expressions that refer to 'main' are a Clang extension">, InGroup<Main>;
```
I think there's work to be done for full conformance here though. For example, this DR clarified that the following code is valid, but we still diagnose it as an extension:
```
int main() {}
decltype(main) maine;
```
https://godbolt.org/z/nn8rd483Y

Comment (Endilll) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: I applied the suggestion.

**Comment by Endilll** [comment]
> do we have test that this is fine?
> 
> ```
> namespace foo {
>     int main();
>     using a = decltype(&main);
> }```
> ```

Thank you, I added the tests.

**Review by cor3ntin — APPROVED** [review]
LGTM if @AaronBallman is happy

("expressions that... are an extensions" still sounds weird to me, cf previous comment)

**Review by AaronBallman — APPROVED** [review]
LGTM!

File: clang/include/clang/Basic/DiagnosticSemaKinds.td (hunk 1 above)
Comment (AaronBallman) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: Good point about `decltype` accepting an expression. This is fine by me.

**Pull request merged**